clang/lib/Sema/SemaConcept.cpp

```diff
@@ -1519,7 +1519,8 @@ bool Sema::IsAtLeastAsConstrained(NamedDecl *D1,
     auto IsExpectedEntity = [](const FunctionDecl *FD) {
       FunctionDecl::TemplatedKind Kind = FD->getTemplatedKind();
       return Kind == FunctionDecl::TK_NonTemplate ||
-             Kind == FunctionDecl::TK_FunctionTemplate;
```

Comment:
So was `TK_FunctionTemplate` never correct? How do we know we have the right set of expected Entities here?

Do the tests we have cover all three cases here? If not can we add a test that covers all three? If not why can't we?

Reply:
@shafik Thanks for taking a look here.

@chaoren Looking at this a bit more closely, I think the suggested removal of `TK_FunctionTemplate` may be incorrect. The affected `Sema::IsAtLeastAsConstrained` function is leveraged by three functions in `SemaTemplateDeduction.cpp`:
- `Sema::getMoreSpecializedTemplate`
- `Sema::getMoreConstrainedFunction`
- `Sema::getMoreSpecialized`

The first of these, `Sema::getMoreSpecializedTemplate`, checks the "constrained-ness" of the templated `FunctionDecl`s associated with two `FunctionTemplateDecl`s. I believe the removed `TK_FunctionTemplate` case would apply here. I'll work backwards from there to see if I can produce a failing test case.

Reply:
@shafik Thanks for taking a look here!

I was able reverse engineer the following program which fails the assertion in question, due to removal of `TK_FunctionTemplate` from the list of "expected" Decls:

```cpp
template <typename T>
struct S {
  template <typename U=T>
    S(U={}) requires (^T != ^void) {}

  template <typename U=T>
    S(U={}) requires (^T != ^bool) {}
};


S<int> s;
```

@cor3ntin As per your original comment, I think this _does_ imply that the assertion is somewhat vacuous - it covers every non-dependent case (well, I guess the `TK_FunctionTemplate` case is sort of dependent too, but whatever).

Would it be better just to remove this assertion altogether, and perhaps to add the above snippet as a test for future reference? Let me know what y'all think would be best.

Reply:
Actually, there might be some deeper issues; will follow up off-thread.